### PR TITLE
Refactor The Resources Page

### LINKED ISSUE #51: (feature): Refactor the Resources page
### Subject

Resources

### Description

Refactor the Resources page.

- Isolate the Talks, Videos and Blogs into their own tabs.
- Add a search filtering functionality inside the tabs.
- ⭐️ Bonus challenge: Utilize a masonry grid to display the resources.

### PR DESCRIPTION
**Pull request by bryg217**
Closes #51 

## 📝 Description

This PR refactors the Resources pages by:

* Isolating the Talks, Videos and Blogs into their own tabs.
* Adding a search filtering functionality inside the tabs.
* Utilizing a masonry grid to display the resources.

## ⛳️ Current behavior (updates)

* Talks, Videos, and Blogs are all visible and not within their own tabs
* No search filtering functionality
* Simple Grid is used for the layout of the resources

## 🚀 New behavior

* Isolate resources: Use Chakra `Tabs`
* Search Filtering: Use Chakra `Input` for element, `formik` for state management, and new utility function for taking in value of input + resources (from props), in order to filter.
  * Filtering Algorithm: Take in query, and for each resource, search its title (i.e. Header) by seeing if that text `includes` query (i.e. `header.includes(query)`).  It's purposefully simple, given the current use case, which is filter a handful of resources on this page, so nothing to comprehensive/complex is needed as of this writing.
* Masonry Grid: New `Masonry` component is used.  More on this below.

## 💣 Is this a breaking change (Yes/No):

No

## 📝 Additional Information

_Note: These are the top items I'm requesting feedback on._

* New Dependencies:
  * `react-responsive-masonry`: Nice little package used to accomplish the masonry layout.  As you'll see below, let me know if there is a simpler/built in way of accomplishing this.  For more on the package, please click [here](https://www.npmjs.com/package/react-responsive-masonry).

Looking forward to some awesome feedback!

Thanks.

### CODE REVIEW

**Comment by bryg217** [comment]
* Looking into resolving merge conflicts ASAP

**Comment by lazarnikolov94** [comment]
@bryg217 [here's a little design](https://www.figma.com/file/Eefs0OXTCUiWroC0P35krU/Chakra-UI-Resources-Page-Redesign?node-id=802%3A9203) I prepared. Basically just wrap the cards with a `VStack` and set the `spacing` to `2`. Make that a `LinkBox` and wrap the title with a `LinkOverlay`. On hover we could scale them by 3-5%. We don't use a lot of shadows in the docsite, but see if it looks okay with a shadow on hover.

**Comment by bryg217** [comment]
@lazarnikolov94 ,

Thanks so much for your support and awesome feedback!  I really appreciate it, since this is my first time contributing to this project.

I'll be implementing and addressing the feedback as soon as I possibly can.

Stay tuned and all the best!

**Comment by bryg217** [comment]
@lazarnikolov94,

Quick update for progress:

* Updated tabs to use `enclosed` variant
* Wrapped cards with `VStack` and `LinkBox`; wrapped header with `LinkOverlay`
* Updated cards padding and background color to reflect the design you sent over
* Spacing at top of tab

Outstanding items I'll be tackling sometime tomorrow:
* On hover scaling
* Box shadow

All the best!

**Comment by lazarnikolov94** [comment]
@bryg217 don't forget the rounded corners! 😁

**Comment by bryg217** [comment]
@lazarnikolov94 ,

Hope all is well!

So, here's the updates I just made:

* Tabs
  * Update `colorScheme` to be `teal`
* Cards
  * Update to use rounded corners
  * On hover
    * Box Shadow
      * This one looks great, but in dark mode, you can't really tell; let me know what you think
    * Scale
      * This one adds a nice touch
      * However, let me know if you pictured it differently; I'd be more than happy to update

As a side note, one thing I did notice was this minor visual bug:

Capture

This happens between `768px` - `770px` resolution, so not too widespread.  However, `768px` is a critical breakpoint, if I am not mistaken.

With all of that being said, any suggestions, from a design standpoint?  (ex: shrinking text, breaking tags to newlines in a certain resolution, etc.).

Looking forward to hearing from you.  All the best.

**Comment by bryg217** [comment]
@lazarnikolov94 ,

In case I don't get to it before you see the updates, I wanted to share one additional update: I had to resolve the conflicts within `yarn.lock`, `package.json`, and others.

So, I did resolve the conflicts.  However, I have to run `yarn lint` again for some of the files I updated, and also do some testing to ensure everything went well, once I get a chance either later tonight or early tomorrow.

Anyway, all the best!

Bryan

**Comment by lazarnikolov94** [comment]
Hey @bryg217! You're almost there! Add some transition animation to the card states:
```typescript
transitionProperty="all"
transitionDuration="slower"
transitionTimingFunction="ease-out"
```

Regarding the tag bleeding out of the card, for that particular case you can break it into two lines:
CleanShot 2021-11-25 at 07 29 15@2x

**Comment by bryg217** [comment]
@lazarnikolov94 ,

Thank you so much for the awesome feedback!  This page is coming together really nicely, thanks to you.  🙏 

So a few updates:

* Added the animation (looks awesome, great idea)
* Properly wrapped the card so that the entire card is clickable, rather than only the header
* Fixed issue where tag was bleeding out of the card
* **Additional and Big Update**: I've installed a new dependency, `minisearch`.  This library is a tiny library that essentially replaces the simple implementation I had before.  Essentially, I was not happy with the results I was getting, using the initial implementation, so I decided to use this, and it's much better.  It uses nice techniques like fuzzy search, ordering based off score, and more.  Anyway, it woks better than I expected it to work , so I am happy with it.  You can view more [here](https://github.com/lucaong/minisearch).  If you think it is overkill, let me know.  We can go back to what we had, but this works really well based off initial testing.

Lastly, I did want to share that I tested on all resolutions (from `360px` and up) on Edge, Chrome, and Firefox.  Based off that, it looks like the implementation is in a great spot.  However, let me know if I missed something!

All the best!

**Comment by lazarnikolov94** [comment]
@bryg217 you've really put in the work! I love it. Although, I think the `minisearch` is an overkill. Let's keep the search only on the title.

**Comment by bryg217** [comment]
@lazarnikolov94,

Thanks so much for the positive feedback, means a lot!  It's been my pleasure having had the opportunity to work on this; I am having a ton of fun!  So, with that being said, I've committed my updates, based on your feedback.

This means:

* Simplify Filtering Implementation
  * It filters by resource title
  * I've removed the custom hook and `minisearch` library

Lastly, I've tested on Firefox, Chrome, and Edge on all resolutions, and it looks great!  However, let me now if you notice something is off.

Thanks!

File: pages/resources.tsx [hunk 1]
```diff
@@ -71,11 +110,24 @@ function ResourceSection(props: ResourceSectionProps) {
...
+        <FormLabel>{t('resources.searchFilter.label')}</FormLabel>
+        <Input
+          onChange={formik.handleChange}
+          placeholder={t('resources.searchFilter.placeholder')}
+          value={formik.values[filterInputId]}
+        />
+        <FormHelperText>{t('resources.searchFilter.helperText')}</FormHelperText>
+      </FormControl>
+      <ResponsiveMasonry columnsCountBreakPoints={{ 350: 1, 580: 2 }}>
+        <Masonry gutter='15px'>
```
Comment (lazarnikolov94) [inline pages/resources.tsx:123]: Let's use values from our design tokens. Even though they're in pixels, make them `16px` so we can have consistent spacing.

**Review by lazarnikolov94 — CHANGES_REQUESTED** [review]

File: pages/resources.tsx (hunk 1 above)
Comment (bryg217) [reply pages/resources.tsx]: Good catch!  I've committed the update, so that now we are using `16px`.

**Comment by lazarnikolov94** [comment]
We should display 1 column on iPad, because the cards look pretty squished.

CleanShot 2021-12-04 at 10 30 50@2x

**Comment by lazarnikolov94** [comment]
Also, let's remove the title label below the tabs, and move the icon in the tab itself. Add some top margin on the tabs as well, it looks pretty close to the description.

**Comment by bryg217** [comment]
@lazarnikolov94 ,

Thanks for the awesome suggestions!

So, quick update:

* I've implemented the second set of updates
  * Removed title label and icon from within the `ResrourceSection`
  * Added icon to tab
  * Added more `margin-top` to the tabs container, so that there is more space between the tabs and the description
* I have **not** implemented:
  * Using 1 column for the tablet; due to some questions I have for that (see below)

Questions for you:

1. Do we want to keep 2 columns for the view ports of `580px` - `767px`, since the side bar is hidden and there seems to be enough space?  Or, do we want it to be 1 column all the way up until desktop?
2. Speaking of desktop, at what point, moving up from tablet, do we want to transition into a 2 column layout?  `850px`?  Some sort of standard desktop breakpoint?
3. I see the card has a max-width of `360px`.  Any specific number you had in mind that we could bump this up to?  Because it might look to small on the one column, when the column itself is over `360px`?

Let me know if you need me to clarify on anything, would be more than happy to.  Thanks for the feedback; this is really coming together nicely!

**Comment by lazarnikolov94** [comment]
Hey @bryg217. You can leave the 2 columns below `768px` since we do have space there. From `768px` until `991px` (which is the `lg` breakpoint), you can set it to 1 column. Anything above should still be 2 columns.

**Comment by bryg217** [comment]
@lazarnikolov94 , alright, it's looking much better now on tablet, and on every other resolution.  Thanks so much for the awesome suggestions!  Let me know what you think.  All the best!

**Review by lazarnikolov94 — APPROVED** [review]
Very nice! 👌

**Comment by lazarnikolov94** [comment]
@allcontributors please add @bryg217 for documentation.

**Comment by bryg217** [comment]
@lazarnikolov94 , thanks so much for the opportunity, I had tons of fun!  Looking forward to future opportunities!  All the best, buddy!  🙏 👍

**Comment by lazarnikolov94** [comment]
> @lazarnikolov94 , thanks so much for the opportunity, I had tons of fun! Looking forward to future opportunities! All the best, buddy! 🙏 👍

Likewise! 😊

**Pull request merged**